contrib/udfs/src/main/java/org/apache/drill/exec/udfs/TimeBucketFunctions.java

```diff
@@ -0,0 +1,105 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.drill.exec.udfs;
+
+import org.apache.drill.exec.expr.DrillSimpleFunc;
+import org.apache.drill.exec.expr.annotations.FunctionTemplate;
+import org.apache.drill.exec.expr.annotations.Output;
+import org.apache.drill.exec.expr.annotations.Param;
+import org.apache.drill.exec.expr.holders.BigIntHolder;
+
+
+public class TimeBucketFunctions {
+
+  /**
+   * This function is used for facilitating time series analysis by creating buckets of time intervals.  See
+   * https://blog.timescale.com/blog/simplified-time-series-analytics-using-the-time_bucket-function/ for usage. The function takes two arguments:
+   * 1. The timestamp in nanoseconds
+   * 2. The desired bucket interval IN MILLISECONDS
+   *
+   * The function returns a BIGINT of the nearest time bucket.
+   */
+  @FunctionTemplate(name = "time_bucket_ns",
+    scope = FunctionTemplate.FunctionScope.SIMPLE,
+    nulls = FunctionTemplate.NullHandling.NULL_IF_NULL)
+  public static class TimeBucketNSFunction implements DrillSimpleFunc {
+
+    @Param
+    BigIntHolder inputDate;
+
+    @Param
+    BigIntHolder interval;
+
+    @Output
+    BigIntHolder out;
+
+    @Override
+    public void setup() {
+    }
+
+    @Override
+    public void eval() {
+      // Get the timestamp in nanoseconds
+      long timestamp = inputDate.value;
+
+      // Get the interval in milliseconds and convert to nanoseconds
+      long intervalToAdd = interval.value * 1000000;
+
+      out.value = (long) java.lang.Math.floor(timestamp / intervalToAdd) * intervalToAdd;
+    }
+  }
+
+  /**
+   * This function is used for facilitating time series analysis by creating buckets of time intervals.  See
+   * https://blog.timescale.com/blog/simplified-time-series-analytics-using-the-time_bucket-function/ for usage. The function takes two arguments:
+   * 1. The timestamp in milliseconds
+   * 2. The desired bucket interval IN milliseconds
+   *
+   * The function returns a BIGINT of the nearest time bucket.
+   */
+  @FunctionTemplate(name = "time_bucket",
+    scope = FunctionTemplate.FunctionScope.SIMPLE,
+    nulls = FunctionTemplate.NullHandling.NULL_IF_NULL)
+  public static class TimeBucketFunction implements DrillSimpleFunc {
+
+    @Param
+    BigIntHolder inputDate;
+
+    @Param
+    BigIntHolder interval;
+
+    @Output
+    BigIntHolder out;
+
+    @Override
+    public void setup() {
+    }
+
+    @Override
+    public void eval() {
+      // Get the timestamp in milliseconds
+      long timestamp = inputDate.value;
+
+      // Get the interval in milliseconds
+      long intervalToAdd = interval.value;
+
+      out.value = (long) java.lang.Math.floor(timestamp / intervalToAdd) * intervalToAdd;
```

Comment:
isn't simpler formula ```timestamp - (timestamp % interval)``` ?

Reply:
@ihuzenko 
Wow... yes.... that is a lot simpler.  That is what happens when you are coding at 2AM... ;-)
(Fixed)

Reply:
no problems, you are a little bit in a rush I guess since in your second commit you left commented line. Also, it would be good to rerun your test just to be sure that the suggestion works fine.

Reply:
@ihuzenko 
I did rerun all the tests before I submitted and everything passed..  But I did forget to remove the commented out line.  I saw that naturally after I pushed, but I got rid of it.  Commits are squashed.  Are we good to go?

Reply:
yes, looks good. Thanks for the changes.

Reply:
@ihuzenko 
Thanks for the review.  Can you (or I) go ahead and commit this to master?  I'm working on a project over the weekend and that would be very helpful.